rq/queue.py

```diff
@@ -337,17 +350,30 @@ def enqueue_dependents(self, job, pipeline=None):
                 pipe.multi()
 
                 for dependent in dependent_jobs:
-                    registry = DeferredJobRegistry(dependent.origin,
-                                                   self.connection,
-                                                   job_class=self.job_class)
-                    registry.remove(dependent, pipeline=pipe)
-                    if dependent.origin == self.name:
-                        self.enqueue_job(dependent, pipeline=pipe)
-                    else:
-                        queue = Queue(name=dependent.origin, connection=self.connection)
-                        queue.enqueue_job(dependent, pipeline=pipe)
-
-                pipe.delete(dependents_key)
+                    dependent.remove_dependency(job.id)
+                    # TODO: pipelining of removal not possible since next if
+                    # expects an empty dependencies key. Is there a better way?
+                    # or is it ok if not everything is pipelined?
+                    if dependent.has_unmet_dependencies() is False:
+                        registry = DeferredJobRegistry(dependent.origin,
+                                                       self.connection,
+                                                       job_class=self.job_class)
+                        registry.remove(dependent, pipeline=pipe)
+                        if dependent.origin == self.name:
+                            self.enqueue_job(dependent, pipeline=pipe)
+                        else:
+                            queue = Queue(name=dependent.origin, connection=self.connection)
+                            queue.enqueue_job(dependent, pipeline=pipe)
```

Comment:
Big problem. It can happen that two workers get to this point at the same time. The consequence is that the same job is enqeued twice. In my system it happened about 1 in 100 times but it still is a problem. I am trying to find a solution but until now this code is not production safe !